### DISCUSSION TITLE
Rate limiting is coming...

### DISCUSSION THREAD

**Discussion by madpilot** [amberelectric/public-api#146] [Announcements]
Hey all,

It's somewhat strange to be excited to introduce a feature that is literally going to limit your usage of our API.

But, we've now got so many users optimising their batteries that a downed API will have a material effect on our users.

You have all been REALLY well-behaved. I can count on one hand the number of times an errant user has taken down the API, and when I've reached out the issues have been fixed really quickly. 

That said, it would be much better for everyone if good behaviour could be self-served, which is why we'll be introducing rate limits. Notice I said _will_. This is an announcement for feedback. Please read on...

# The greatest hits

* We'll be implementing [the draft IETF rate limit header proposal](https://datatracker.ietf.org/doc/draft-ietf-httpapi-ratelimit-headers/) so you'll see new headers returned that will give you an idea of what your limit is, and when it will reset. **This is implemented _now_** so you can see how close to the sun you are flying, _without_ the limit being enforced. We are also monitor for people that hit the limits, so we can tune the numbers before we start enforcing the limits. [The docs](https://app.amber.com.au/developers/documentation/) have already been updated.
* We've started with a limit of 50 calls per 5-minutes (reset every 5-minutes). Given that our prices really only update once every 5-minutes, that should give you support for 50 devices. Realistically, people seem to poll every minute, which is still 10 devices. As I mentioned, we will be reviewing the numbers before we flick the switch. We can also bump up the numbers on a customer-by-customer basis if there is a good reason.
* API limits are per account, not per site, or per API key. Again, if that is super limiting, we'll review before we go live properly.

Please let me know whether those numbers feel workable.

Cheers!

**Comment by shadow7412** [comment]
10 calls per minute sounds like it shouldn't be an issue for sane api users. I don't see any detriments.

**Comment by ps-jay** [comment]
TIL there's a draft RFC for rate limiting! ..  cool

**Comment by freman** [comment]
Seems perfectly reasonable to me, wouldn't take but a moment to knock out a caching proxy for people that have 100 devices 🤣

**Comment by chiefcomm** [comment]
Seems logical to me

**Comment by wh0phd** [comment]
@madpilot Myles, was this the subject of the banner announcement on the app tonight (Tue 15 Aug 2023) ?

If so, I can help answer "Using time based control in the Tesla app *may* work" — there was a mystery export limit between 6:05pm and 6:31pm tonight, stopping me from discharging my 2x PowerWalls more than 5kW total.  When I called Tesla support, they confirmed (but I don't want to take this as 100%) that both batteries were working and each was discharging 2.5kW.  This barely covered the oven operation and left very little for the windfall prices happening at 6:00pm-6:30pm tonight.

(I initially blamed the induction stove installation last week, which has to be on non-backed-up B phase, but it wasn't running tonight)

Here is the email I asked them to send me, noting their phone conversation with me.

> This is Jose Mora from the Powerwall Support team in Energy Customer Support. Thank you so much for reaching out.
> 
> From the review here it looks like the batteries may have an issue with the discharge rate. If the issue occurred after the install  of the device a few days ago I would recommend a review of the install. I will also escalate this case to our Tier 2 Team for review.
> 
> Please let us know if you have anymore questions or concerns.

`Powerwall Inquiry Activity - 50809426 [refid:CI00BDC8AC]`

**Reply by madpilot** [reply]
No, this banner is about our public API.

Last night was Tesla rate limiting us.

**Reply by madpilot** [reply]
And, Thanks for the extra intel!
